doc/neps/nep-0055-string_dtype.rst

```diff
@@ -656,13 +518,385 @@ string validation in a separate, expensive, pass over the input array-like. We
+opaque ``npy_string_allocator`` struct. Internally, it has the following layout:
+
+.. code-block:: c
+
+    struct npy_string_allocator {
+        npy_string_malloc_func malloc;
+        npy_string_free_func free;
+        npy_string_realloc_func realloc;
+        npy_string_arena arena;
+    };
+
+This allows us to group memory-allocation functions together and choose
+different allocation functions at runtime if we desire.  Use of
+the allocator is guarded by a mutex, see below for more discussion about thread
+safety.
+
+The memory allocations are handled bit the ``npy_string_arena`` struct member,
+which has the following layout:
+
+.. code-block:: c
+
+    struct npy_string_arena {
+        size_t cursor;
+        size_t size;
+        char *buffer;
+    };
+
+Where ``buffer`` is a pointer to the beginning of a heap-allocated arena,
+``size`` is the size of that allocation, and ``cursor`` is the location in the
+arena where the last arena allocation ended. The arena is filled using an
+exponentially expanding buffer, allowing amortized O(1) insertion.
+
+Each string entry in the arena is prepended by a size, stored either in a
+``char`` or a ``size_t``, depending on the length of the string. Strings with
+lengths between 16 or 8 (depending on architecture) and 255 are stored with a
+``char`` size. We refer to these as "medium" strings internally and strings
+stored this way have the ``NPY_STRING_MEDIUM`` flag set. This choice reduces the
+overhead for storing smaller strings on the heap by 7 bytes per medium-length
+string.
+
+The size of the allocation is stored in the arena to allow reuse of the arena
+allocation if a string is mutated. In principle we could disallow re-use of the
+arena buffer and not store the sizes in the arena. This may or may not save
+memory or be more performant depending on the exact usage pattern. For now we
+are erring on the side of avoiding unnecessary heap allocations when a string is
+mutated but in principle we could simplify the implementation by choosing to
+always store mutated arena strings as heap strings and ignore the arena
+allocation. See more details below on how we deal with the mutability of NumPy
+arrays in a multithreaded context.
+
+Using a per-array arena allocator ensures that the string buffers for nearby
+array elements are usually nearby on the heap. We do not guarantee that
+neighboring array elements are contiguous on the heap to support the small
+string optimization, missing data, and allow mutation of array entries. See
+below for more discussion on how these topics affect the memory layout.
+
+If the contents of a packed string are freed and then assigned to a new string
+with the same size or smaller than the string that was originally stored in the
+packed string, the existing short string or arena allocation is re-used, with
+padding zeros written to the end of the subset of the buffer reserved for the
```

Comment:
I don't think we have to zero pad?  Aalthough I am totally fine with doing so.

Reply:
No, we don't have to zero-pad. I figure it's bad to leave behind stale data. We can always decide not to do this later if it's a performance concern.